Packet++/src/TcpSorter.cpp

```diff
@@ -0,0 +1,585 @@
+#define LOG_MODULE PacketLogModuleTcpSorter
+
+#include "TcpSorter.h"
+#include "TcpLayer.h"
+#include "IPv4Layer.h"
+#include "IPv6Layer.h"
+#include "PacketUtils.h"
+#include "IpAddress.h"
+#include "Logger.h"
+#include <sstream>
+#include <random>
+#include <vector>
+#include <algorithm>
+#if defined(WIN32) || defined(PCAPPP_MINGW_ENV) //for using ntohl, ntohs, etc.
+#include <winsock2.h>
+#elif LINUX
+#include <in.h> //for using ntohl, ntohs, etc.
+#elif MAC_OS_X || FREEBSD
+#include <arpa/inet.h> //for using ntohl, ntohs, etc.
+#endif
+
+#define PURGE_FREQ_SECS 1
+
+namespace pcpp
+{
+
+TcpSorter::TcpSorter(
+							OnTcpPacketReady onPacketReadyCallback,
+							OnTcpPacketMissing onPacketMissingCallback,
+							void* userCookie,
+							const TcpSorterConfiguration &config)
+{
+	m_isClosed = false;
+	m_OnPacketReadyCallback = onPacketReadyCallback;
+	m_OnPacketMissingCallback = onPacketMissingCallback;
+	m_UserCookie = userCookie;
+	// configuration value
+	m_MaxNumCapturedPacket = config.maxNumCapturedPacket;
+	m_MaxIdleTimeout = config.maxIdleTimeout;
+	m_MaxNumInactiveConnScan = config.maxNumInactiveConnScan;
+	m_MaxSegmentLifeTime = config.maxSegmentLifeTime;
+	m_CleanUpInactiveConnPeriod = config.cleanUpInactiveConnPeriod;
+	m_ShouldIncludeEmptySegments = config.shouldIncludeEmptySegments;
+	// reset clean up timer
+	m_LastCleanupTime = time(NULL);
+}
+
+TcpSorter::~TcpSorter()
+{
+	closeAllConnections();
+}
+
+void TcpSorter::sortPacket(SPRawPacket spRawPacket)
+{
+	// stop sorting if it is closed
+	if(m_isClosed)
+	{
+		return;
+	}
+	// parse the packet
+	Packet packet(spRawPacket.get(), false);
+
+	time_t now = time(NULL);
+
+	// clean up any expired inactive TCP connection
+	if (0 != m_MaxIdleTimeout && m_LastCleanupTime + m_CleanUpInactiveConnPeriod < now)
+	{
+		cleanUpInactiveTcpConnection(now);
+		m_LastCleanupTime = now;
+	}
+
+	// get IP layer
+	Layer* ipLayer = nullptr;
+	if (packet.isPacketOfType(IPv4))
+		ipLayer = dynamic_cast<Layer*>(packet.getLayerOfType<IPv4Layer>());
+	else if (packet.isPacketOfType(IPv6))
+		ipLayer = dynamic_cast<Layer*>(packet.getLayerOfType<IPv6Layer>());
+
+	if (ipLayer == nullptr)
+		return;
+
+	// Ignore non-TCP packets
+	TcpLayer* tcpLayer = packet.getLayerOfType<TcpLayer>();
+	if (tcpLayer == nullptr)
+		return;
```

Comment:
Possible optimizations:

```c++
  // get IP layer
  Layer* ipLayer;
  if (packet.isPacketOfType(IPv4))
    ipLayer = dynamic_cast<Layer*>(packet.getLayerOfType<IPv4Layer>());
  else if (packet.isPacketOfType(IPv6))
    ipLayer = dynamic_cast<Layer*>(packet.getLayerOfType<IPv6Layer>());
  else
    return;

  // Ignore non-TCP packets
  TcpLayer* tcpLayer = packet.getNextLayerOfType<TcpLayer>(ipLayer);
  if (tcpLayer == nullptr)
    return;
```

If you are sure that the packet is parsed until TCP the following code can be used:

```c++
  // get IP layer
  Layer* ipLayer;
  if (packet.isPacketOfType(IPv4))
    ipLayer = dynamic_cast<Layer*>(packet.getLayerOfType<IPv4Layer>());
  else if (packet.isPacketOfType(IPv6))
    ipLayer = dynamic_cast<Layer*>(packet.getLayerOfType<IPv6Layer>());
  else
    return;

  // Ignore non-TCP packets
  if (packet.isPacketOfType(TCP))
    return;
	
  TcpLayer* tcpLayer = packet.getLastLayer();
```